dao/src/main/java/com/epam/esm/dao/impl/TagDaoImpl.java

```diff
@@ -0,0 +1,84 @@
+package com.epam.esm.dao.impl;
+
+import com.epam.esm.dao.TagDao;
+import com.epam.esm.dao.exception.WrongInsertDataException;
+import com.epam.esm.domain.entity.Tag;
+import com.epam.esm.dao.mapper.TagMapper;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Component;
+
+import java.sql.PreparedStatement;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+@Component
+public class TagDaoImpl implements TagDao {
+
+    private static final String SELECT_TAG_BY_NAME = "SELECT * FROM tag WHERE name=?";
+    private final String SELECT_TAGS_FOR_CERTIFICATE = "SELECT t.id, t.name FROM tag AS t " +
+            "INNER JOIN gift_certificate_has_tag AS ght ON t.id = ght.tag_id " +
+            "INNER JOIN gift_certificate AS g ON g.id=ght.gift_certificate_id " +
+            "WHERE gift_certificate_id=?";
+    private static final String CREATE_TAG = "INSERT INTO tag (`name`) SELECT ? " +
+            "WHERE NOT EXISTS (SELECT name FROM tag WHERE name=? LIMIT 1)";
+    private static final String SELECT_TAG_BY_ID = "SELECT * FROM tag WHERE id=?";
+    private static final String SELECT_ALL_TAGS = "SELECT * FROM tag";
+    private static final String DELETE_TAG = "DELETE FROM tag WHERE tag.id = ? AND NOT EXISTS " +
+            "(SELECT * FROM gift_certificate_has_tag AS ght WHERE ght.tag_id = ?)";
+    private static final String COLUMN_ID = "id";

```

Comment:
I believe it's the same "id" as ID in CertificateMapper. If so, it look like a duplication. The same idea for tags.

Reply:
all db column names moved to ColumnName class to avoid code duplication